Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp

```diff
@@ -597,7 +598,13 @@ class TIFFLoadingContext {
             return read_tiff_value(type, count, offset);
         }()));
 
-        TRY(handle_tag(m_metadata, tag, type, count, move(tiff_value)));
+        auto subifd_handler = [&](u32 ifd_offset) -> ErrorOr<void> {
+            m_next_ifd = ifd_offset;
+            TRY(read_next_image_file_directory());
```

Comment:
Another question, sorry. `read_next_idf_offset()` calls `read_next_idf_offset()` at the end, to read linked IDFs at the current level. But where do we call `read_next_image_file_directory()` for that? (Also for the current code, not just for the code new in this PR.)

Reply:
We don't yet!
But that will probably be something like
```cpp
do {
    read_next_ifd();
} while (m_next_ifd.has_value());
```
In `read_header()`.

While this works to read all IFD, everything is overwritten so that's not really useful.

Reply:
…edit: Ignore approval then :D

Reply:
Actually, have you seen any file containing these following sublevels ifds? I don't think I have...

Reply:
Yes, see I think they're pretty common. e.g. https://github.com/nico/jpegfiles/blob/main/DJI_0406.JPG:

```
% ~/src/hack/jpeg_exif_dump DJI_0406.JPG 
ffe1 at offset 2, size 29962: Application Segment (APP1)
  app id: 'Exif'
...
  tag 34665 (Exif IFD) format 4 (unsigned long): count 1: 182
  tag 34853 (GPSInfo) format 4 (unsigned long): count 1: 686
...
  exif IFD
    tag 33434 (ExposureTime (seconds)) format 5 (unsigned rational): count 1: 1/640 (0.002)
...
    tag 40965 (InteroperabilityIFD) format 4 (unsigned long): count 1: 656
...
    interoperability IFD:
      tag 1 (InteroperabilityIndex) format 2 (ascii): count 4: 'R98'
...
  GPSInfo IFD:
    tag 0 format 1 (unsigned byte): count 4
...
  next IFD at 776
  tag 259 (Compression) format 3 (unsigned short): count 1: 6
...
```

I think the 776 is the next IFD of the outer, global exif IFD (while `exif IFD` is stored in 34665 in the global exif, and the interop IFD is stored in there in tag 40965, and GPS is also stored in 34853 in the global exif).

(One thing that's also interesting here is that it shows that exif tags aren't global but namespaced: The interop IFD has its own tag namespace, and the GPS IFD does too.)

Reply:
There is a misunderstanding, so let's backtrack a little bit.
IFD are composed of tags and a pointer to next IFD. Following the pointer will be called a horizontal move. Also, some tags can contain pointer to other IFD, these are always related to the parent IFD. In the file you provided, it is the case of the Exif IFD at offset 182. Following them will be called a vertical move.

This PR adds support for vertical moves, and your question is about horizontal ones. The patch in the previous comment adds support for horizontal moves, but only if they are at the top level. And my comment was that I've never seen a child IFD having siblings, in other words horizontal moves when depth > 0.

The IFD at offset 776 is, AFAICT, simply another image in this TIFF container (encoder usually encode thumbnails this way).


Output of our decoder with the patch mentioned earlier:
<details>
```
UnknownTag(270, ASCII): DCIM\100MEDIA\DJI_0406.JPG
Make(ASCII): DJI                          
Model(ASCII): FC7303                       
Orientation(UnsignedShort): Default
XResolution(UnsignedRational): 72 (72/1)
YResolution(UnsignedRational): 72 (72/1)
ResolutionUnit(UnsignedShort): Inch
Software(ASCII): v01.43.0062            
DateTime(ASCII): 2022:04:27 15:41:59
UnknownTag(531, UnsignedShort): 1
ExifIFD(UnsignedLong): 182
Reading image file directory at offset 182
UnknownTag(33434, UnsignedRational): 0.0015625 (1/640)
UnknownTag(33437, UnsignedRational): 2.8 (280/100)
UnknownTag(34850, UnsignedShort): 2
UnknownTag(34855, UnsignedShort): 100
UnknownTag(36864, Undefined): Buffer of size: 4
UnknownTag(36867, ASCII): 2022:04:27 15:41:59
UnknownTag(36868, ASCII): 2022:04:27 15:41:59
UnknownTag(37121, Undefined): Buffer of size: 4
UnknownTag(37122, UnsignedRational): 3.8341933333333333 (5751290/1500000)
UnknownTag(37377, SignedRational): 9.3219 (93219/10000)
UnknownTag(37378, UnsignedRational): 2.97 (297/100)
UnknownTag(37380, SignedRational): 0 (0/3)
UnknownTag(37381, UnsignedRational): 2.97 (297/100)
UnknownTag(37382, UnsignedRational): 0 (0/100)
UnknownTag(37383, UnsignedShort): 2
UnknownTag(37384, UnsignedShort): 1
UnknownTag(37385, UnsignedShort): 32
UnknownTag(37386, UnsignedRational): 4.49 (449/100)
UnknownTag(37500, Undefined): Buffer of size: 16384
UnknownTag(40960, Undefined): Buffer of size: 4
UnknownTag(40961, UnsignedShort): 1
UnknownTag(40962, UnsignedLong): 4000
UnknownTag(40963, UnsignedLong): 3000
UnknownTag(40965, UnsignedLong): 656
UnknownTag(41493, UnsignedRational): nan (0/0)
UnknownTag(41728, Undefined): Buffer of size: 1
UnknownTag(41729, Undefined): Buffer of size: 1
UnknownTag(41985, UnsignedShort): 0
UnknownTag(41986, UnsignedShort): 0
UnknownTag(41987, UnsignedShort): 0
UnknownTag(41988, UnsignedRational): 1 (100/100)
UnknownTag(41989, UnsignedShort): 24
UnknownTag(41990, UnsignedShort): 0
UnknownTag(41991, UnsignedShort): 0
UnknownTag(41992, UnsignedShort): 0
UnknownTag(41993, UnsignedShort): 0
UnknownTag(41994, UnsignedShort): 0
UnknownTag(41996, UnsignedShort): 0
UnknownTag(42033, ASCII): 1SFOJ8F0AB01G3                                 
UnknownTag(34853, UnsignedLong): 686
UnknownTag(40092, Byte): Buffer of size: 256
UnknownTag(40094, Byte): Buffer of size: 256
Reading image file directory at offset 776
Compression(UnsignedShort): OldJPEG
XResolution(UnsignedRational): 72 (72/1)
YResolution(UnsignedRational): 72 (72/1)
ResolutionUnit(UnsignedShort): Inch
UnknownTag(513, UnsignedLong): 18420
UnknownTag(514, UnsignedLong): 11534
```
</details>

Reply:
I understand the horizontal / vertical bits and that this patch adds the latter, and that my question is about the former :)

Aha, I haven't seen horizontal moves at depth > 0 either. I have seen it at depth == 0 (see example). I did misunderstand your question.

(In addition to thumbnails, the level 0 exif chain can also be used for the MPF extension I think.)

Reply:
Sorry if I sounded like I was explaining to you what you said. I was trying to remove confusion by properly defining my words and link them to your ideas.

We're not ready to support IFD in a chain at depth 0 for other reasons, and as it is not common to find these chains at depth > 0, let's punt that for later us :)